sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -684,11 +686,40 @@ public String toString() {
       return MoreObjects.toStringHelper(Write.class).add("config", getBigtableConfig()).toString();
     }
 
+    private static class SeenWindow {
+      private BoundedWindow window;
+      private Instant timestamp;
+
+      public SeenWindow(BoundedWindow window, Instant timestamp) {
+        this.window = window;
+        this.timestamp = timestamp;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        return (o instanceof SeenWindow) && ((SeenWindow) o).window.equals(window);
+      }
+
+      @Override
+      public int hashCode() {
+        return window.hashCode();
+      }
+
+      public BoundedWindow getWindow() {
+        return window;
+      }
+
+      public Instant getTimestamp() {
+        return timestamp;
+      }
+    }
+
     private class BigtableWriterFn extends DoFn<KV<ByteString, Iterable<Mutation>>, Void> {
 
       public BigtableWriterFn(BigtableConfig bigtableConfig) {
         this.config = bigtableConfig;
         this.failures = new ConcurrentLinkedQueue<>();
+        this.seenWindows = new HashSet<>();
```

Comment:
Seems like this would be much cleaner in terms of Java readability as a simple `HashMap<BoundedWindow, Instant>`.

However, looking at code below, you just need a `HashSet<BoundedWindow>`, and then use window.maxTimestamp() instead of the timestamp of a non-deterministically chosen element within the window.

Reply:
hmm, yeah that seems reasonable!